mlir/include/mlir/Dialect/MIOpen/utility/loweringUtils.h

```diff
@@ -22,27 +22,59 @@
 namespace mlir {
 namespace miopen {
 
-inline int64_t calculateKBlockNum(int64_t n, int64_t ho, int64_t wo) {
-  int64_t gemmK = n * ho * wo;
-  int64_t gemmKBlocks = 1;
-  if (gemmK % 16 == 0) {
-    auto lcm = math_util::lcm(ho * wo, (int64_t)16);
-    gemmKBlocks = std::min(gemmK / lcm, n);
-  } else if (gemmK % 8 == 0) {
-    auto comm = math_util::lcm(ho * wo, (int64_t)8);
-    gemmKBlocks = std::min(gemmK / comm, n);
-  } else if (gemmK % 4 == 0) {
-    auto comm = math_util::lcm(ho * wo, (int64_t)4);
-    gemmKBlocks = std::min(gemmK / comm, n);
+// Heuristic logic to compute KBlock for backward weight atomic add kernel.
+// The logic is adopted from MIOpen.
+//
+// The logic searches within the range of [1, 20 * number of CUs / gridSize],
+// where gridSize is the original number of workgroups required for the
+// convolution, and find the largest KBlock number which preserves the 2
+// contraints:
+// - GemmK (before splitting) = KBlock * KPerBlock * KPack * GemmK (after
+// splitting).
+// - n (batch size) is divisible by KBlock.
+//
+// 20 is a magic number obtained in MIOpen after empirical testing. It offers a
+// reasonable reduction of GemmK after splitting, without incurring too much
+// overheads on atomic adds. One potential future work is to make this value be
+// tunable.
+inline LogicalResult calculateKBlockNum(int64_t n, int64_t ho, int64_t wo,
```

Comment:
I'll abstain from reviewing this function. Still unclear about how line 59-69 works.